Add mass and time unit to new project wizard panel

falken10vdl opened the pull request:
This is a proposal to address issue https://github.com/IfcOpenShell/IfcOpenShell/issues/7073
Cheers!

falken10vdl committed:
proposal to add massunit and timeunit in project creation wizard

falken10vdl committed:
use kilogram as default for SI and Seconds for Imperial

Andrej730 commented:
@falken10vdl 

Looks good in general, a few things I've noticed.

1. Some Bonsai tests are now failing.

2. `add_conversion_based_unit` - `if unit_type == "MASSUNIT":` this will never be true for `ton` units, since they're not present in `imperial_types`. I think we can just add it to that dictionary, but probably should spell it as "tonne" - it seems to be the established way to spell "metric ton" https://en.wikipedia.org/wiki/Tonne

3. Need to add tests for new behaviour in `add_conversion_based_unit`.

5. `unit.assign_unit` - I'm not sure if we should introduce any arguments and code to handle them to `assign_unit`, since it's basically going to duplicate code from `add_si_unit` / `add_conversion_based_unit` and we were planning to deprecate existing `length`, `area`, `volume` arguments.

falken10vdl committed:
Updated with core developer feedback: fixed for Tonne and added tests

falken10vdl commented:
@Andrej730 Thanks a lot for the feedback. I have changed the error by adding TONNE as suggested. I have also added tests.
I add them here how to run them for my own reference ;) 


```
falken10vdl@monster:~/bonsaiDevel/IfcOpenShell/src/ifcopenshell-python$ python -m pytest test/util/test_unit.py -v
Blender 4.5.2 LTS (hash ab25eae04993 built 2025-08-20 02:11:19)
module changed on disk: '/home/falken10vdl/.config/blender/4.5/extensions/blender_org/sun_position/__init__.py' reloading...

----------------------------------------------

No module named 'blenderbim'
Archipack ifc export fallback
Archipack PRO 2.8.2 : ready 

----------------------------------------------

🔄 blenderkit: Verbose is enabled
🔄 blenderkit: Read in JSON settings from file
ℹ️  blenderkit: Ports reordered so first port is now 62485 (previous index was 0) [20:21:44.910, client_lib.py:125]
Created history step c5fc304c-c138-49b8-96a0-eb97658ac250
ℹ️  blenderkit: Created new search history step [20:21:44.938, search.py:1736]
========================================================================================================================================== test session starts ==========================================================================================================================================
platform linux -- Python 3.11.11, pytest-8.4.1, pluggy-1.6.0 -- /home/falken10vdl/.local/share/applications/blender-4.5.2-linux-x64/4.5/python/bin/python3.11
cachedir: .pytest_cache
rootdir: /home/falken10vdl/bonsaiDevel/IfcOpenShell/src/ifcopenshell-python
configfile: pyproject.toml
plugins: bdd-8.1.0
collected 38 items                                                                                                                                                                                                                                                                                      

test/util/test_unit.py::TestCacheUnits::test_run PASSED                                                                                                                                                                                                                                           [  2%]
test/util/test_unit.py::TestClearUnitCache::test_run PASSED                                                                                                                                                                                                                                       [  5%]
test/util/test_unit.py::TestGetProjectUnit::test_run PASSED                                                                                                                                                                                                                                       [  7%]
test/util/test_unit.py::TestGetProjectUnit::test_using_a_cache PASSED                                                                                                                                                                                                                             [ 10%]
test/util/test_unit.py::TestGetPropertyUnit::test_no_unit PASSED                                                                                                                                                                                                                                  [ 13%]
test/util/test_unit.py::TestGetPropertyUnit::test_simple_quantity PASSED                                                                                                                                                                                                                          [ 15%]
test/util/test_unit.py::TestGetPropertyUnit::test_single_value PASSED                                                                                                                                                                                                                             [ 18%]
test/util/test_unit.py::TestGetPropertyUnit::test_enumerated_value PASSED                                                                                                                                                                                                                         [ 21%]
test/util/test_unit.py::TestGetPropertyUnit::test_list_value PASSED                                                                                                                                                                                                                               [ 23%]
test/util/test_unit.py::TestGetPropertyUnit::test_bounded_value PASSED                                                                                                                                                                                                                            [ 26%]
test/util/test_unit.py::TestConvert::test_run PASSED                                                                                                                                                                                                                                              [ 28%]
test/util/test_unit.py::TestCalculateUnitScale::test_prefix_and_conversion_based_units_are_considered PASSED                                                                                                                                                                                      [ 31%]
test/util/test_unit.py::TestFormatLength::test_run PASSED                                                                                                                                                                                                                                         [ 34%]
test/util/test_unit.py::TestIsAttrType::test_run PASSED                                                                                                                                                                                                                                           [ 36%]
test/util/test_unit.py::TestConvertFileLengthUnits::test_run PASSED                                                                                                                                                                                                                               [ 39%]
test/util/test_unit.py::TestConvertFileLengthUnits::test_attribute_conversion PASSED                                                                                                                                                                                                              [ 42%]
test/util/test_unit.py::TestConvertFileLengthUnits::test_converting_map_conversion_if_there_is_no_map_unit PASSED                                                                                                                                                                                 [ 44%]
test/util/test_unit.py::TestConvertFileLengthUnits::test_preserving_enh_if_there_is_a_map_unit PASSED                                                                                                                                                                                             [ 47%]
test/util/test_unit.py::TestConvertFileLengthUnits::test_preserving_enh_if_there_is_a_map_unit_which_is_also_the_project_default PASSED                                                                                                                                                           [ 50%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4::test_run PASSED                                                                                                                                                                                                                           [ 52%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4::test_attribute_conversion PASSED                                                                                                                                                                                                          [ 55%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4::test_converting_map_conversion_if_there_is_no_map_unit PASSED                                                                                                                                                                             [ 57%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4::test_preserving_enh_if_there_is_a_map_unit PASSED                                                                                                                                                                                         [ 60%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4::test_preserving_enh_if_there_is_a_map_unit_which_is_also_the_project_default PASSED                                                                                                                                                       [ 63%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4X3::test_run PASSED                                                                                                                                                                                                                         [ 65%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4X3::test_attribute_conversion PASSED                                                                                                                                                                                                        [ 68%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4X3::test_converting_map_conversion_if_there_is_no_map_unit PASSED                                                                                                                                                                           [ 71%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4X3::test_preserving_enh_if_there_is_a_map_unit PASSED                                                                                                                                                                                       [ 73%]
test/util/test_unit.py::TestConvertFileLengthUnitsIFC4X3::test_preserving_enh_if_there_is_a_map_unit_which_is_also_the_project_default PASSED                                                                                                                                                     [ 76%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_adding_mass_units_creates_proper_massunit PASSED                                                                                                                                                                              [ 78%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_adding_time_units_creates_proper_timeunit PASSED                                                                                                                                                                              [ 81%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_mass_unit_integration_with_project PASSED                                                                                                                                                                                     [ 84%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_time_unit_integration_with_project PASSED                                                                                                                                                                                     [ 86%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_multiple_unit_types_can_coexist PASSED                                                                                                                                                                                        [ 89%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_unknown_mass_or_time_units_fall_back_to_userdefined PASSED                                                                                                                                                                    [ 92%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_mass_units_individually PASSED                                                                                                                                                                                                [ 94%]
test/util/test_unit.py::TestAddConversionBasedUnitMassAndTime::test_time_units_individually PASSED                                                                                                                                                                                                [ 97%]
test/util/test_unit.py::TestMassAndTimeUnitDictionaries::test_mass_units_in_imperial_types PASSED                                                                                                                                                                                                 [100%]

========================================================================================================================================== 38 passed in 0.16s ===========================================================================================================================================
ℹ️  blenderkit: Unregistering BlenderKit add-on [20:21:45.628, __init__.py:2553]
falken10vdl@monster:~/bonsaiDevel/IfcOpenShell/src/ifcopenshell-python$ cd -
/home/falken10vdl/bonsaiDevel/IfcOpenShell/src/bonsai
falken10vdl@monster:~/bonsaiDevel/IfcOpenShell/src/bonsai$ pytest -p no:pytest-blender test/core/test_unit.py 
========================================================================================================================================== test session starts ==========================================================================================================================================
platform linux -- Python 3.11.13, pytest-8.4.1, pluggy-1.6.0
rootdir: /home/falken10vdl/bonsaiDevel/IfcOpenShell/src/bonsai
configfile: pytest.ini
plugins: blender-3.0.7, bdd-8.1.0
collected 30 items                                                                                                                                                                                                                                                                                      

test/core/test_unit.py ..............................                                                                                                                                                                                                                                             [100%]

========================================================================================================================================== 30 passed in 0.03s ===========================================================================================================================================
falken10vdl@monster:~/bonsaiDevel/IfcOpenShell/src/bonsai$ python -m pytest test/tool/test_unit.py
Blender 4.5.2 LTS (hash ab25eae04993 built 2025-08-20 02:11:19)
module changed on disk: '/home/falken10vdl/.config/blender/4.5/extensions/blender_org/sun_position/__init__.py' reloading...

----------------------------------------------

No module named 'blenderbim'
Archipack ifc export fallback
Archipack PRO 2.8.2 : ready 

----------------------------------------------

🔄 blenderkit: Verbose is enabled
🔄 blenderkit: Read in JSON settings from file
ℹ️  blenderkit: Ports reordered so first port is now 62485 (previous index was 0) [20:22:08.426, client_lib.py:125]
Created history step 30aaa3b4-eb65-444b-acc7-8eaaf7d2f141
ℹ️  blenderkit: Created new search history step [20:22:08.455, search.py:1736]
========================================================================================================================================== test session starts ==========================================================================================================================================
platform linux -- Python 3.11.11, pytest-8.4.1, pluggy-1.6.0
rootdir: /home/falken10vdl/bonsaiDevel/IfcOpenShell/src/bonsai
configfile: pytest.ini
plugins: blender-3.0.7, bdd-8.1.0
collected 29 items                                                                                                                                                                                                                                                                                      

test/tool/test_unit.py .............................                                                                                                                                                                                                                                              [100%]

========================================================================================================================================== 29 passed in 0.75s ===========================================================================================================================================
ℹ️  blenderkit: Unregistering BlenderKit add-on [20:22:09.735, __init__.py:2553]
falken10vdl@monster:~/bonsaiDevel/IfcOpenShell/src/bonsai$ ^C
falken10vdl@monster:~/bonsaiDevel/IfcOpenShell/src/bonsai$ 


```


> unit.assign_unit - I'm not sure if we should introduce any arguments and code to handle them to assign_unit, since it's basically going to duplicate code from add_si_unit / add_conversion_based_unit and we were planning to deprecate existing length, area, volume arguments.


I just followed the pattern for  `length`, `area`, `volume` arguments. What would you suggest to do?

Thanks!



```

Andrej730 commented:
@falken10vdl 

1. `src/ifcopenshell-python/test/util/test_unit.py` - it's a wrong place for `add_conversion_based_unit` tests, we typically use `api` package for api tests - https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifcopenshell-python/test/api/unit/test_add_conversion_based_unit.py

2. > I just followed the pattern for length, area, volume arguments. What would you suggest to do?

    To avoid adding new arguments, since we plan to deprecate previous ones.

3. `TestMassAndTimeUnitDictionaries` - there's misindentation, so `test_mass_units_in_imperial_types` and other methods are considered to be local to `test_mass_units_in_imperial_types` and never actually executed.

    And what's the idea for those tests? It seems they're just rechecking the values of the constants that already defined in the dictionaries.

4. `test_mass_unit_integration_with_project`, `test_time_unit_integration_with_project`, `test_multiple_unit_types_can_coexist` - those tests seems to be unrelated to `add_conversion_based_unit`.

5. `test_unknown_mass_or_time_units_fall_back_to_userdefined` - why checking it twice if there's no distinction between mass and time units in this case?

6. `test_mass_units_individually`, `test_time_units_individually` - those do exactly the same as `test_adding_mass_units_creates_proper_massunit` and `test_adding_time_units_creates_proper_timeunit`.

falken10vdl committed:
Updated with core developer feedback

falken10vdl commented:
@Andrej730 

>    src/ifcopenshell-python/test/util/test_unit.py - it's a wrong place for add_conversion_based_unit tests, we typically use api package for api tests - https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifcopenshell-python/test/api/unit/test_add_conversion_based_unit.py

ok removed form test_unit.py and moved to test_add_conversion_based_unit.py

>    To avoid adding new arguments, since we plan to deprecate previous ones.

I have removed the parameters

>    TestMassAndTimeUnitDictionaries - there's misindentation, so test_mass_units_in_imperial_types and other methods are considered to be local to test_mass_units_in_imperial_types and never actually executed.  And what's the idea for those tests? It seems they're just rechecking the values of the constants that already defined in the dictionaries.

Yes, those follow under the category of "testing constants". I agree that probably do not make much sense since the dictionary team and the testing team are the same developers :) I can remove them if that is creating confusion

>    test_mass_unit_integration_with_project, test_time_unit_integration_with_project, test_multiple_unit_types_can_coexist - those tests seems to be unrelated to add_conversion_based_unit.

Yes I removed them (may be more related to assing_unit)

>    test_unknown_mass_or_time_units_fall_back_to_userdefined - why checking it twice if there's no distinction between mass and time units in this case?

Agree. Reduced to one

>    test_mass_units_individually, test_time_units_individually - those do exactly the same as test_adding_mass_units_creates_proper_massunit and test_adding_time_units_creates_proper_timeunit.

Agree. Reduced

Andrej730 commented:
@falken10vdl 

>moved to test_add_conversion_based_unit.py

Tests are added only to IFC4 test class, they should be added to IFC2X3 test class so they would be executed for both IFC2X3 and IFC4.

> I have removed the parameters

They're still present in `assign_unit` doc-string.

> Yes, those follow under the category of "testing constants". I agree that probably do not make much sense since the dictionary team and the testing team are the same developers :) I can remove them if that is creating confusion

I don't think we need them. Also, those constants are already checked in `test_adding_mass_units_creates_proper_massunit`, `test_adding_time_units_creates_proper_timeunit`.

falken10vdl committed:
Cleanup based on core developer's feedback

falken10vdl commented:
@Andrej730 

> Tests are added only to IFC4 test class, they should be added to IFC2X3 test class so they would be executed for both IFC2X3 and IFC4.

Done

> They're still present in assign_unit doc-string.

I just put an example but not provided as explicit parameter

> I don't think we need them. Also, those constants are already checked in test_adding_mass_units_creates_proper_massunit, test_adding_time_units_creates_proper_timeunit.

Removed

falken10vdl committed:
black formatting

Moult commented:
Is it worth having this in the wizard? Mass and time are not used by everyone. It's not hard to add later on in the units panel.

falken10vdl commented:
As you wish. I think this was brought by  @steverugi which is something useful for Costing and Scheduling. Maybe moving it to that panel  would make sense?
Cheers!

steverugi commented:
@Moult , @falken10vdl 
I agree both are not indispensable, just convenient for those who need them, if a problem to have them as default they can be added later

falken10vdl commented:
@Moult @steverugi  Maybe an option is to just add a toggle to add or remove mass&time units. By defualt it is off.
Something like this:

https://github.com/user-attachments/assets/513f4d0a-6bcd-4e1e-afdf-d3daeb14cd52

Cheers!

steverugi commented:
> @Moult @steverugi Maybe an option is to just add a toggle to add or remove mass&time units. By defualt it is off. Something like this:
> 
>  
> Cheers!

Love it! thanks

falken10vdl committed:
ui change to add or not add mass and time units

Andrej730 commented:
I think a button for this draws too much attention from "Create Project" button.

Maybe foldable panel (https://docs.blender.org/api/current/bpy.types.UILayout.html#bpy.types.UILayout.panel) will work for this better?

steverugi commented:
Hi @Andrej730 
I'll be happy with the collapsable panel as you suggest
thanks a lot

falken10vdl commented:
@Andrej730 & @steverugi Something like this?
<img width="774" height="755" alt="image" src="https://github.com/user-attachments/assets/0f534e56-acac-419d-a803-d6c936e19337" />

And then:
<img width="654" height="825" alt="image" src="https://github.com/user-attachments/assets/31cb2ec0-d6e1-4c11-bd25-212ef67ab9c1" />
<img width="657" height="867" alt="image" src="https://github.com/user-attachments/assets/780f04ad-afff-44e6-9a0e-b4d8df799311" />

Cheers!

falken10vdl committed:
Add mass and time units as a panel to project wizard UI

steverugi commented:
I am not sure this is going to be acceptable but I wouldn't mind having all options in a drop-down menu, saved for Schema, System, and Length
Which means in the drop-down menu you can have Area, Volume, Mass, Time together.

Area, Volume, and Mass by default follow the system (square/cubic foot or meter, pound or kg) while Time is day for both.

Typically you choose Metric or Imperial, sometimes you change length, the rest is pretty much standard.

Thanks again for your (Much appreciated) assistance!

falken10vdl committed:
Refactor mass and time units panel into collapsible section using .panel method  in project wizard UI

falken10vdl commented:
Sorry I was using a whole subpanel. Here as proposed by @Andrej730 just to use the.panel method in the UILayout:
<img width="655" height="883" alt="image" src="https://github.com/user-attachments/assets/077d5a3a-f8ca-4f6b-afbe-a9863083078f" />

Much cleaner.
Cheers!

Andrej730 commented:
Yeah, much clearer. I was thinking maybe there's no need for this "Add Units" button at all? Just add those units anyway - so users wouldn't need to think about what "add / remove units" would do.

Actually I also liked @steverugi idea - most of the time users need just a length unit to start a project, so maybe we can also move Area and Volume to this "Additional Units" tab. Thoughts? @Moult @theoryshaw 

PS Thinking about the better ways to hide those additional units from Project Wizard, makes me think that maybe they should be only editable after project was created... Though then users won't be able to setup a default unit settings in their startup blend file, so maybe keeping them kind of hidden could be a compromise.

theoryshaw commented:
i think listing them all right out of the gate is the best way. 
....

A user can set those settings they use a lot, and save it to the 'startup file', and that way they never have to set them again when they create a new .ifc file.


https://github.com/user-attachments/assets/23dfa4fb-dcd9-4865-a511-041dc716cf9f

falken10vdl commented:
The idea of having the button was to either insert them or remove them. From @Moult comments above I had the impression that many users (I guess most) do not necessarily use the mass or time units and having in them the units added to the file does not make too much sense if they are never going to be used.

I think that a panel for "additional units" would make sense and as @steverugi was pointing out, may just leave Schema and length unit out and the rest under "additional units".

By the way, why there is "Unit System"? I guess if you chose a length of the type feet, miles, etc it is implied you want "Imperial" and therefore the display of units follow Imperial. I do not see anything in the ifc file that records that "unit system" or I am missing something?

Cheers!

falken10vdl commented:
I have just created a new project with "Imperial" but squared meters in area unit:
<img width="648" height="450" alt="image" src="https://github.com/user-attachments/assets/55e4b518-88e7-4f31-9477-1cbad92169cf" />

and I see the following:
<img width="2187" height="1739" alt="image" src="https://github.com/user-attachments/assets/f449db05-475b-4d98-b1d4-124b73220a1a" />

So it looks that the measure tool uses the length as base and the area is just that one squared while the quantity take off is following the units assignment. In this case I had created a slab of 10 feet length and that is 9.29 sqared meters or 100 squared feet.
Should I rise an issue about the measure tool to be consistent with the area unit selected? (and I guess area unit should be just area length squared if it is not defined by the user?)
Cheers!

falken10vdl committed:
Add preference for mass and time units visibility in project wizard

falken10vdl commented:
@steverugi @theoryshaw @Andrej730 I have added this as an "extra feature" so users can enable or disable at will (extras by default come disabled to maintain previous behavior).

<img width="1581" height="705" alt="image" src="https://github.com/user-attachments/assets/40670504-67ae-48fa-b913-db2e9e1a3ed5" />

Cheers!
